### PR TITLE
Give the users customizable axis label limits (Fixes #2085)

### LINKED ISSUE #2085: How to set LabelCount bigger than 25 in HorizatalBarView

### PR DESCRIPTION
**Pull request by igzrobertoestrada**
This change adds the ability to users to override the imposed limitation of at least 2 labels per axis and at most 25 labels per axis if they need to and at their own risk. 

By default these customizable limits are hardcoded to the previous limits, at least 2 labels, at most 25.

The rationale for this change is that when coming from previous versions of Charts, users are able to use –at least on the X axis– as many labels as they wish and when they upgrade to a newer version of Charts they find that they have lost functionality, and this change brings that functionality back.

### CODE REVIEW

**Comment by liuxuan30** [comment]
I am not sure if allowing other values could raise any issue when >25 ?

**Comment by igzrobertoestrada** [comment]
I'm using 61 labels in the X axis in an AppStore production app and it's working just fine. 

In my case, this value comes as a client requirement for a feature. Many other people may be in the same or similar situation just as they tell in #2085 and they and/or their clients cannot afford a step back in functionality compared to what they could have in prior versions.

Many people will be safe with the 25-labels limitation. It will work wonders for them and they won't even notice it, but others may need more flexibility when it comes to customizing their charts in order to satisfy requirements that –in many cases– are non negotiable.

**Comment by liuxuan30** [comment]
I wonder how can you draw 61 labels on x axis on a mobile screen? ipad or iPhone? I am OK with this change. But what I concerned is in somewhere, the is one piece of code that only thinks about <25 values, if >25, a bug triggered.
@danielgindi do you remember anything? If this is safe to merge?

**Comment by alejandroluengo** [comment]
It's ok, merge needed for sure

**Comment by igzrobertoestrada** [comment]
@liuxuan30 I have the barchart inside a UIScrollView so the users can swipe through the chart because having the entire chart on a relatively small region of an iPhone screen in portrait mode would be an awful user experience and really really difficult to read.

**Comment by liuxuan30** [comment]
@igzrobertoestrada why your question is related to this PR?
It's your choice to show the full chart or zoom in to only show some; and there are APIs to restrict how many items you can see. e.g. visibleXrange.

**Comment by igzrobertoestrada** [comment]
@liuxuan30 Because I need to maintain a piece of code that requires that charts is able to render more than 25 labels as it did in previous versions.

File: Source/Charts/Components/AxisBase.swift [hunk 1]
```diff
@@ -213,6 +213,16 @@ open class AxisBase: ComponentBase
     /// the total range of values this axis covers
     @objc open var axisRange = Double(0)
     
+    /// The minumum number of labels on the axis
+    @objc open var axisMinLabels: Int = Int(2) {
```
Comment (jjatie) [inline Source/Charts/Components/AxisBase.swift]: Type inference is sufficient. `@objc open var axisMinLabels = 2`

**Comment by jjatie** [comment]
@igzrobertoestrada Isn't having the chart in a `UIScrollView` with many labels, the same things as just scrolling through a chart with fewer labels? The only difference is that in the latter case the you can always see the y-axis.

File: Source/Charts/Components/AxisBase.swift (hunk 1 above)
Comment (igzrobertoestrada) [reply Source/Charts/Components/AxisBase.swift]: Fixed in new commit. Removed ```: Int``` but maintained ```Int(2)``` for consistence with the rest of the codebase.

File: Source/Charts/Components/AxisBase.swift [hunk 2]
```diff
@@ -214,12 +214,12 @@ open class AxisBase: ComponentBase
     @objc open var axisRange = Double(0)
     
     /// The minumum number of labels on the axis
-    @objc open var axisMinLabels: Int = Int(2) {
+    @objc open var axisMinLabels = Int(2) {
         didSet { axisMinLabels = axisMinLabels > 0 ? axisMinLabels : oldValue }
```
Comment (liuxuan30) [inline Source/Charts/Components/AxisBase.swift]: should we use `didSet` or setter for this check here? @jjatie

**Comment by liuxuan30** [comment]
Again I don't have issue to let user decide the value instead o capping it at 25. What concerns me is I don't know if there is any function that take 25 as granted, when we change to larger ones, it broke. Only @danielgindi might remember.
If we think this can be merged and later let user report what's breaking, it's good to go

File: Source/Charts/Components/AxisBase.swift [hunk 3]
```diff
@@ -213,6 +213,16 @@ open class AxisBase: ComponentBase
...
+        didSet { axisMinLabels = axisMinLabels > 0 ? axisMinLabels : oldValue }
+    }
+    
+    /// The maximum number of labels on the axis
+    @objc open var axisMaxLabels = Int(25) {
+        didSet { axisMinLabels = axisMaxLabels > 0 ? axisMaxLabels : oldValue }
+    }
+    
     /// the number of label entries the axis should have
     /// max = 25,
```
Comment (liuxuan30) [inline Source/Charts/Components/AxisBase.swift]: comment might consider to change as well.

**Comment by jjatie** [comment]
@danielgindi What are your thoughts on adding this to the library? Technically there's no reason to impose a limit on the number of labels, but from a data visualization perspective it makes sense to limit it.

**Comment by jjatie** [comment]
@liuxuan30 I think if there are functions that use require 25 already, we should consider those bugs that need to be fixed. If we accept this change, we will just have to fix those issues as they come up.

**Comment by liuxuan30** [comment]
I just dont know if there is any. Just my guess. So far I dont see

**Comment by liuxuan30** [comment]
let's merge it

**Comment by jjatie** [comment]
@liuxuan30 FYI I'm available for about an hour.

**Pull request merged**